### PR TITLE
Refactor ops and fn APIs

### PR DESCRIPTION
**Pull request by klecki**
## Description: 
This part of refactoring moves the most important parts of handling the argument
parsing, deprecation, constant promotion, conditionals, etc to the _OperatorInstance.
In case of ops API, there needs to be a pre-step that merges the state
from `__init__` and `__call__` and is specific that API.
For the `fn` API, apart from promoting constants first (due to the perf reasons),
we can handle the creation of `_OperatorInstance` and filling the `spec` in one go.

`name` is no longer included in argument inputs but considered scalar argument.

It tries to keep most of the implementation in one place, but also allows for detaching
the ops and fn APIs from one another.
The question is, how much of this change we want.

Additionally, some details around the conditionals are streamlined, allowing
to enable them in multiple input sets.

Next steps:
Python function, TFRecord reader and external source implement their Operator class
without inheriting from the one generated in op factory, so they repeat some code.
We probably can refactor it and unify them with the generic code path even more.
Make the special arguments (name, device), less special so they are added to the
spec through generic code-path.

### Affected modules and functionalities:
fn, ops

### Key points relevant for the review:
Should we do the thing that is put in the fn part as a comment?

### Tests:

- [x] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/ops/__init__.py [hunk 1]
```diff
@@ -206,75 +311,67 @@ def _add_spec_args(schema, spec, kwargs):
...
+    the C++ backend.
 
-    def __init__(self, inputs, op, **kwargs):
+    Parameters
+    ----------
+    object : _type_
+        _description_
+    """
+
+    def __init__(self, inputs, arguments, arg_inputs, op):
```
Comment (klecki) [inline dali/python/nvidia/dali/ops/__init__.py:330]: Consider propagating the arguments, that were already added to the spec in __init__ for full visibility.

**Review by szalpal — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/ops/__init__.py [hunk 2]
```diff
@@ -189,6 +189,116 @@ def _handle_deprecations(schema, kwargs, op_name):
     return kwargs
 
 
+def _handle_op_deprecation(schema, op_name):
+    if schema.IsDeprecated():
+        # TODO(klecki): how to know if this is fn or ops?
```
Comment (stiepan) [inline dali/python/nvidia/dali/ops/__init__.py:194]: So for the class-based API we're going to suggest `fn.*` to now? Was it like that before?

File: dali/python/nvidia/dali/ops/__init__.py [hunk 3]
```diff
@@ -189,6 +189,116 @@ def _handle_deprecations(schema, kwargs, op_name):
...
+        with warnings.catch_warnings():
+            warnings.simplefilter("default")
+            warnings.warn(msg, DeprecationWarning, stacklevel=2)
+
+
+def _resolve_double_definitions(current, previous, keep_old=True):
+    """Unify the arguments (or argument inputs) between __call__ and __init__ - ops API only.
+
+    If the same argument was passed in both invocation it should result in error, with exception,
+    of passing `None` in `__call__`.
```
Comment (stiepan) [inline dali/python/nvidia/dali/ops/__init__.py:211]:
```suggestion
    If the same argument was passed in both invocation it should result in error, with exception
    of passing `None` in `__call__`.
```

File: dali/python/nvidia/dali/ops/__init__.py (hunk 2 above)
Comment (klecki) [reply dali/python/nvidia/dali/ops/__init__.py:194]: Yup, this was just moved around.
I think the best way would be to add some `_api_type` string and dispatch on that in errors (also propagate it to the backend as hidden arg possibly. But this is a case for follow-up.

**Review by stiepan — APPROVED** [review]

**Pull request merged**